frontend/app/model/MembershipCatalog.scala

```diff
@@ -0,0 +1,219 @@
+    _backendType: BackendType
+  ): MembershipCatalog = new MembershipCatalog {
+
+    override def staff = _staff
+    override def patron = _patron
+    override def partner = _partner
+    override def friend = _friend
+    override def supporter = _supporter
+    override def backendType = _backendType
+    override def tierPlanDetails(ratePlanId: ProductRatePlanId): Option[TierPlanDetails] = tierPlanDetailsMap.get(ratePlanId)
+  }
+
+  def fromZuora(ratePlanIds: RatePlanIds)(ratePlans: Seq[rest.ProductRatePlan])(implicit bt: BackendType): Val[MembershipCatalog] = {
+    val ratePlansById = ratePlans.map(p => p.id -> p).toMap
+    def byId(tierPlan: TierPlan, id: ProductRatePlanId): Val[ProductRatePlan] =
+      ratePlansById
+        .get(id)
+        .map(_.successNel)
+        .getOrElse((tierPlan -> s"""Failed to find the plan with id "$id"""").failureNel)
+
+    def planCharge(plan: ProductRatePlan) = plan.productRatePlanCharges match {
+      case Seq(c) => c.successNel
+      case Nil => s"Could not find any rate plan charge for ${plan.id}".failureNel
+      case cs => s"Ambiguous rate plan charges found for ${plan.id}: ${cs.mkString(", ")}".failureNel
+    }
+
+    def planAndCharge(tierPlan: TierPlan, ratePlanId: String): Val[(ProductRatePlan, ProductRatePlanCharge)] =
+      byId(tierPlan, ratePlanId).flatMap { plan =>
+        planCharge(plan).leftMap(_.map(tierPlan -> _)).map {charge =>
+          (plan, charge)
+        }
+      }
+
+    def freeTierPlanDetails(tierPlan: FreeTierPlan, ratePlanId: ProductRatePlanId): Val[FreeTierPlanDetails] =
+      byId(tierPlan, ratePlanId).map { plan =>
+        FreeTierPlanDetails(tierPlan, plan.id)
+      }
+
+    def paidTierPlanDetails(tierPlan: PaidTierPlan, ratePlanId: ProductRatePlanId): Val[PaidTierPlanDetails] =
+      planAndCharge(tierPlan, ratePlanId).flatMap { case (plan, charge) =>
+        val summary: Either[NonEmptyList[Err], PricingSummary] =
+          PricingSummary.fromRatePlanCharge(charge)
+            .left.map(_.map(tierPlan -> _).toList.toNel.get)
+
+        Validation.fromEither(summary)
+          .map { summary =>
+            PaidTierPlanDetails(tierPlan, plan.id, summary)
+          }
+      }
+
+    def currenciesMismatchErr(tierDetails: PaidTierDetails): NonEmptyList[Err] = {
+      val msg = "The supported currencies for the different plans do not match"
+      NonEmptyList(
+        tierDetails.monthlyPlanDetails.plan -> msg,
+        tierDetails.yearlyPlanDetails.plan -> msg
+      )
+    }
+
+    def currenciesMatch(tierDetails: PaidTierDetails): Boolean =
+      tierDetails.monthlyPlanDetails.currencies == tierDetails.yearlyPlanDetails.currencies
+
+    object PublicPlans {
+      val ids = ratePlanIds
+      val friend = freeTierPlanDetails(FriendTierPlan.public, ids.friend)
+      val staff = freeTierPlanDetails(StaffPlan, ids.staff)
+      val supporterM = paidTierPlanDetails(PaidTierPlan.monthly(Supporter, Public), ids.supporterMonthly)
```

Comment:
I think you can reduce the number of declared variables here and just go with loops to reduce repetition.